site/en/blog/crx-scripting-api/index.md

```diff
@@ -0,0 +1,306 @@
+
+### Injecting a function with arguments
+
+While considering how the platform would need to evolve in light of remotely-hosted code
+restrictions, we wanted to find a balance between the raw power of arbitrary code execution and only
+allowing static content scripts. The solution we struck on was to allow extensions to inject a
+function as a content script and to pass an array of values as arguments.
+
+Let's take a quick look at an (oversimplified) example. Say we wanted to inject a script that
+greeted the user by name when the user clicks the extension's action button (icon in the toolbar).
+In Manifest V2, we could dynamically construct a code string and execute that script in the current
+page.
+
+```js
+// Manifest V2 extension
+chrome.browserAction.onClicked.addListener(async (tab) => {
+  let userReq = await fetch('https://example.com/greet-user.js');
+  let userScript = await userReq.text();
+
+  chrome.tabs.executeScript({
+    // userScript == 'alert("Hello, <GIVEN_NAME>!")'
+    code: userScript,
+  });
+});
+
+```
+
+While Manifest V3 extensions can't use code that isn't bundled with the extension, our goal was to
+preserve some of the dynamism that arbitrary code blocks enabled for Manifest V2 extensions. The
+function and arguments approach makes it possible for Chrome Web Store reviewers, users, and other
+interested parties to more accurately assess the risks an extension poses while also allowing
+developers to modify an extension's runtime behavior based on user settings or application state.
+
+```js
+// Manifest V3 extension
+function greetUser(name) {
+  alert(`Hello, ${name}!`);
+}
+chrome.action.onClicked.addListener(async(tab) => {
+  let userReq = await fetch('/https://example.com/user-data.json');
+  let user = await userReq.json();
+  let givenName = user.givenName || '<GIVEN_NAME>';
+
+  chrome.scripting.executeScript({
+    target: { tabId: tab.id },
+    func: greetUser,
+    args: [givenName],
+  });
+});
+```
+
+### Targeting frames
+
+We also wanted to improve how developers interact with frames in the revised API. The Manifest V2
+version of `executeScript` allowed developers to target either all frames in a tab or a specific
+frame in the tab.
+
+```js
+// Manifest V2 extension
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.webNavigation.getAllFrames({tabId: tab.id}, (frames) => {
+    let frame1 = frames[0].frameId;
+    let frame2 = frames[1].frameId;
+
+    chrome.tabs.executeScript(tab.id, {
+      frameId: frame1,
+      file: 'content-script.js',
+    });
+    chrome.tabs.executeScript(tab.id, {
+      frameId: frame2,
+      file: 'content-script.js',
+    });
+  });
+});
+```
+
+In Manifest V3, we replaced the optional `frameId` integer property in the options object with an
+optional `frameIds` array of integers; this allows developers to target multiple frames in a single
+API call. You can use use `chrome.webNavigation.getAllFrames` to get a list of all frames in a tab.
```

Comment:
nit: put the part about getAllFrames() above, since we use it in the MV2 example, too

Reply:
Done